Elytron OIDC client

jmesnil opened the pull request:
* Update the example instructions
* Put the application at the ROOT of the server
* Simplify the Helm configuration

Signed-off-by: Jeff Mesnil <jmesnil@redhat.com>

jmesnil commented:
@fjuma Could you please review? 

I tried to deploy on OpenShift Dev Sandbox with RH SSO 7.5 and the example failed until I added the `http://` to the `Valid Redirect URIs`. Otherwise, I had an error: `Invalid parameter: redirect_uri` with the query parameter set to `redirect_uri: http://elytron-oidc-client-app-jmesnil1-dev.apps.sandbox.x8i5.p1.openshiftapps.com/secured`

This seems wrong to me that the app that I publicly access on the `https://` would contact RH SSO with a `redirect_uri` using HTTP instead of HTTPS.

@fjuma do you have any idea which configuration I'm missing to make sure that the app is setting the proper redirect_uri so that I don't have to list the `http://` URL in the valid redirect URIs?

jmesnil commented:
@fjuma, FIY, I've opened https://issues.redhat.com/browse/ELY-2389 to improve the expression resolution of Elytron so that the oidc.json could be configured with 

```json 
"provider-url": "${oidc.provider-url}",
```

and we would be able to use either the `oidc.provider-url` System property or the `OIDC_PROVIDER_URL` env var to configure it.

This would be consistent with how WildFly and MicroProfile can resolve properties.

fjuma commented:
@jmesnil I'll take a closer look at the questions here in a bit but just wanted to mention it's already possible to use an environment variable in an `oidc.json` file (similar to the way it used to work with a `keycloak.json` file). As an example, it can be configured using `env.OIDC_PROVIDER_URL` as shown below:

"provider-url" : "${env.OIDC_PROVIDER_URL:http://localhost:8080}/realms/myrealm",

jmesnil commented:
> 'll take a closer look at the questions here in a bit but just wanted to mention it's already possible to use an environment variable in an oidc.json file

That's what I'm doing in my PR, thanks :) Now, I also want to remove the `env.` prefix ;)

fjuma commented:
> That's what I'm doing in my PR, thanks :) Now, I also want to remove the `env.` prefix ;)

Ah, ok, sorry, got it now :) Was reading the comment quickly but I see the details in the issue now. Thanks for creating that!

jfdenise commented on examples/elytron-oidc-client/helm.yaml:
```diff
@@ -1,11 +1,10 @@
 build:
-  uri: https://github.com/wildfly/wildfly-s2i
-  mode: s2i
+  uri: https://github.com/jmesnil/wildfly-s2i
```
Seems that you are using your own fork.

jmesnil replied:
fixed, thanks

jfdenise commented on examples/elytron-oidc-client/helm.yaml:
```diff
@@ -1,11 +1,10 @@
-    version: latest
-    builderImage: quay.io/wildfly/wildfly-s2i-jdk11
-    runtimeImage: quay.io/wildfly/wildfly-runtime-jdk11
+  ref: elytron-oidc-client
```
Fork branch.

fjuma commented:
Just FYI, I've pinged Pedro Silva from the Keycloak team to ask about the `redirect_uri` issue on OpenShift. Locally, things work fine with HTTPS so there must be some configuration we are missing on OpenShift.

jfdenise commented:
@fjuma , I tried to investigate the https problem. It seems that the elytron-oidc-client sends the redirect URL when interacting with the SSO server. Am I correct? I can see in the SSO server traces:  
```
10:34:31,259 WARN [org.keycloak.events] (default task-14) type=LOGIN_ERROR, realmId=WildFly, clientId=simple-webapp, userId=null, ipAddress=88.171.142.107, error=invalid_redirect_uri, redirect_uri=http://elytron-oidc-client-app-jdenise-dev.apps.sandbox-m2.ll9k.p1.openshiftapps.com/simple-webapp/secured
```
although the client configured in the SSO server contains https redirect URL, we can notice that the server receies http redirect url. It seems that it only does that when using oidc.json.

When using the auto-registration of client (that configures the subsystem and create the client on the server), the https redirect URI works correctly. An example of subsystem configuration that we generate at startup:
```
<subsystem xmlns="urn:wildfly:elytron-oidc-client:1.0">
            <provider name="keycloak">
                <provider-url>https://sso-jdenise-dev.apps.sandbox-m2.ll9k.p1.openshiftapps.com/auth/realms/WildFly</provider-url>
                <ssl-required>external</ssl-required>
                <allow-any-hostname>false</allow-any-hostname>
                <disable-trust-manager>true</disable-trust-manager>
                <enable-cors>false</enable-cors>
                <register-node-at-startup>true</register-node-at-startup>
                <register-node-period>600</register-node-period>
            </provider>
            <secure-deployment name="ROOT.war">
                <provider>keycloak</provider>
                <client-id>root</client-id>
                <bearer-only>false</bearer-only>
                <enable-basic-auth>true</enable-basic-auth>
                <credential name="secret" secret="mysecret"/>
            </secure-deployment>
        </subsystem>
```
In this case, the redirect URL are https, and the subsystem send https url when interacting with the server (I configured the SSO client with invalid https URL and I can see the following error:

```
11:47:48,021 WARN [org.keycloak.events] (default task-13) type=LOGIN_ERROR, realmId=WildFly, clientId=root, userId=null, ipAddress=88.171.142.107, error=invalid_redirect_uri, redirect_uri=https://elytron-oidc-client-app-auto-reg-jdenise-dev.apps.sandbox-m2.ll9k.p1.openshiftapps.com/secured
``` 
You can notice the `https://elytron-oidc-client-app-auto-reg-jdenise...`

How the subsystem findout what are the redirect URL to send?

Are we missing something in the following oidc.json?
```
{
  "provider-url": "${org.wildfly.s2i.example.oidc.provider-url}",
  "ssl-required": "external",
  "allow-any-hostname": true,
  "disable-trust-manager": true,
  "client-id": "simple-webapp",
  "public-client": true,
  "confidential-port": 0
}
```

jfdenise commented:
@fjuma , yes, redirect to https. You can look at the auto-registration example there: https://github.com/wildfly/wildfly-s2i/tree/main/examples/elytron-oidc-client-auto-reg

fjuma commented:
@jfdenise The code that creates the `redirectUri` is here:

https://github.com/wildfly-security/wildfly-elytron/blob/1.x/http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcAuthenticationMechanism.java#L60

However, this ultimately just relies on the actual request URL that was passed when calling `OidcAuthenticationMechanism#evaluateRequest`, see:

https://github.com/wildfly-security/wildfly-elytron/blob/1.x/http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcRequestAuthenticator.java#L150

and

https://github.com/wildfly-security/wildfly-elytron/blob/1.x/http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcAuthenticationMechanism.java#L60

When debugging, we can see that for the auto registration case, the request coming in has HTTPS (e.g., check the value of `request.getRequestUri()`) but for the other case, the request coming in does not have HTTPS. It's still unclear to me what causes us to receive different requests in these two cases.

jfdenise commented:
@fjuma , I investigated more and found the root cause. We are not setting proxy-address-forwarding=true that is required to use the initial URL. We are doing it for the auto-registration case (thanks to using the cloud feature-pack).
@jmesnil , you should add the cloud FP in the pom.xml, then https redirect will work.
Thank-you for your help @fjuma!

fjuma commented:
@jfdenise Ah, that's good to know.

jmesnil commented:
@jfdenise thanks for the investigation.

Once I added the `wildfly-cloud-galleon-pack`, there was no issue with the HTTPS URL.
I've fixed the changes you requested too.
